group/edwards25519/scalar.go

```diff
@@ -4,11 +4,196 @@
 // from SUPERCOP.
 
 // The scalars are GF(2^252 + 27742317777372353535851937790883648493).
 
+type scalar struct {
+	v [32]byte
+}
+
+// Equality test for two Scalars derived from the same Group
+func (s *scalar) Equal(s2 kyber.Scalar) bool {
+	v1 := s.v[:]
+	v2 := s2.(*scalar).v[:]
+	return subtle.ConstantTimeCompare(v1, v2) != 0
+}
+
+// Set equal to another Scalar a
+func (s *scalar) Set(a kyber.Scalar) kyber.Scalar {
+	s.v = a.(*scalar).v
+	return s
+}
+
+func (s *scalar) Clone() kyber.Scalar {
+	s2 := *s
+	return &s2
+}
+
+func (s *scalar) setInt(i *mod.Int) kyber.Scalar {
+	b := i.LittleEndian(32, 32)
+	copy(s.v[:], b)
+	return s
+}
+
+// Set to a small integer value
+func (s *scalar) SetInt64(v int64) kyber.Scalar {
+	return s.setInt(mod.NewInt64(v, primeOrder))
+}
+
+func (s *scalar) toInt() *mod.Int {
+	return mod.NewIntBytes(s.v[:], primeOrder, mod.LittleEndian)
+}
+
+// Set to the additive identity (0)
+func (s *scalar) Zero() kyber.Scalar {
+	s.v = [32]byte{0}
+	return s
+}
+
+// Set to the multiplicative identity (1)
+func (s *scalar) One() kyber.Scalar {
+	s.v = [32]byte{1}
+	return s
+}
+
+// Set to the modular sum of scalars a and b
+func (s *scalar) Add(a, b kyber.Scalar) kyber.Scalar {
+	scAdd(&s.v, &a.(*scalar).v, &b.(*scalar).v)
+	return s
+}
+
+// Set to the modular difference a - b
+func (s *scalar) Sub(a, b kyber.Scalar) kyber.Scalar {
+	scSub(&s.v, &a.(*scalar).v, &b.(*scalar).v)
+	return s
+}
+
+// Set to the modular negation of scalar a
+func (s *scalar) Neg(a kyber.Scalar) kyber.Scalar {
+	var z scalar
+	z.Zero()
+	scSub(&s.v, &z.v, &a.(*scalar).v)
+	return s
+}
+
+// Set to the modular product of scalars a and b
+func (s *scalar) Mul(a, b kyber.Scalar) kyber.Scalar {
+	scMul(&s.v, &a.(*scalar).v, &b.(*scalar).v)
+	return s
+}
+
+// Set to the modular division of scalar a by scalar b
+func (s *scalar) Div(a, b kyber.Scalar) kyber.Scalar {
+	var i scalar
+	i.Inv(b)
+	scMul(&s.v, &a.(*scalar).v, &i.v)
+	return s
+}
+
+// Set to the modular inverse of scalar a
+func (s *scalar) Inv(a kyber.Scalar) kyber.Scalar {
+	var res scalar
+	res.One()
+	ac := a.(*scalar)
+	// Modular inversion in a multiplicative group is a^(phi(m)-1) = a^-1 mod m
+	// Since m is prime, phi(m) = m - 1 => a^(m-2) = a^-1 mod m.
+	// The inverse is computed  using the exponentation-and-square algorithm.
+	// Implementation is constant time regarding the value a, it only depends on
+	// the modulo.
+	for i := 255; i >= 0; i-- {
+		bit := l_minus_2_big.Bit(i)
+		// square step
+		scMul(&res.v, &res.v, &res.v)
+		if bit == 1 {
+			// multiply step
+			scMul(&res.v, &res.v, &ac.v)
+		}
+	}
+	s.v = res.v
+	return s
+}
+
+// Set to a fresh random or pseudo-random scalar
+func (s *scalar) Pick(rand cipher.Stream) kyber.Scalar {
+	i := mod.NewInt(random.Int(primeOrder, rand), primeOrder)
+	return s.setInt(i)
+}
+
+func (s *scalar) SetBytes(b []byte) kyber.Scalar {
+	// XXX handle simple and scReduce cases appropriately
```

Comment:
`scReduce` takes only 64 bytes as input, so I'm not sure we should use it when given more than 64 bytes in `SetBytes`. Otherwise, we need to cut down the given slice of bytes and that may yield wrong results ?. i would keep it the way it is now. What do you think ?